### PR TITLE
Run web API test suite in GitHub Actions

### PR DESCRIPTION
**Pull request by nucleogenic**
Added Pytest API test suite to GitHub Actions web workflow.

**Example run:**
* https://github.com/akuker/RASCSI/actions/runs/3570177332

# Changes

**Refactored Docker infrastructure:**

- Moved various files/directories (sorry for the messy diff!)
- Added healthchecks to backend and web containers
- Reduced Docker image sizes
- Fixed ignore patterns in .dockerignore
- Various other adjustments (e.g. entrypoints)

**Misc:**

- Removed RaSCSI references in various areas (e.g. rascsi -> backend)
- Added compilation-only step to easyinstall.sh
- Moved apt package lists to variables in easyinstall.sh
- Revert to triggering GitHub Actions runs on push as this works better for forks
- Added logging of Protobuf commands
- Added logging of HTTP requests
- Fixed bug in supported browser detection for requests when an absent/unrecognized user agent
- Updated tests to prefix file names with the test name

### CODE REVIEW

**Comment by nucleogenic** [comment]
Note: This should be merged after PR #1010 as the web/frontend_checks paths have been updated to match their new locations in that branch.

File: .dockerignore [hunk 1]
```diff
@@ -2,9 +2,8 @@
 /*
 
 # Paths to include
-!/docker/rascsi/rascsi_wrapper.sh
-!/docker/rascsi/cfilesystem.patch
-!/docker/rascsi-web/start.sh
+!/docker/backend/rascsi_wrapper.sh
```
Comment (rdmark) [inline .dockerignore:5]: Maybe simply name this `wrapper.sh`? Or `backend_wrapper.sh`?

File: docker/README.md [hunk 2]
```diff
@@ -35,14 +35,12 @@ from another terminal.
...
 | `OS_VERSION`         | buster   |
-| `OS_ARCH`            | amd64    |
 | `WEB_HTTP_PORT`      | 8080     |
 | `WEB_HTTPS_PORT`     | 8443     |
 | `WEB_LOG_LEVEL`      | info     |
-| `RASCSI_HOST`        | rascsi   |
+| `RASCSI_HOST`        | backend  |
 | `RASCSI_PORT`        | 6868     |
 | `RASCSI_PASSWORD`    | *[None]* |
 | `RASCSI_LOG_LEVEL`   | debug    |
```
Comment (rdmark) [inline docker/README.md:46]: Argument for removing the `RASCSI` prefix for the env variables.

File: easyinstall.sh [hunk 3]
```diff
@@ -69,6 +69,11 @@ SECRET_FILE="$HOME/.config/rascsi/rascsi_secret"
 FILE_SHARE_PATH="$HOME/shared_files"
 FILE_SHARE_NAME="Pi File Server"
 
+APT_PACKAGES_COMMON="build-essential git protobuf-compiler bridge-utils"
+APT_PACKAGES_BACKEND="libspdlog-dev libpcap-dev libprotobuf-dev protobuf-compiler libgmock-dev clang-11"
+APT_PACKAGES_PYTHON="python3 python3-dev python3-pip python3-venv python3-setuptools python3-wheel libev-dev libevdev2"
+APT_PACKAGES_WEB="nginx-light genisoimage man2html hfsutils dosfstools kpartx unzip unar disktype"
```
Comment (rdmark) [inline easyinstall.sh:75]: This is so much cleaner. :)

File: python/common/src/rascsi/file_cmds.py [hunk 4]
```diff
@@ -47,6 +47,13 @@ def __init__(self, sock_cmd: SocketCmds, ractl: RaCtlCmds, token=None, locale=No
         self.token = token
         self.locale = locale
 
+    def send_pb_command(self, command):
+        if logging.getLogger().isEnabledFor(logging.DEBUG):
+            # TODO: Uncouple/move to common dependency
```
Comment (rdmark) [inline python/common/src/rascsi/file_cmds.py:52]: Can you please elaborate on what you had in mind for a better place to put this pb command logging?

Comment (rdmark) [reply python/common/src/rascsi/file_cmds.py:52]: Ah I see, you repeat the same method in the other classes.

File: python/web/src/templates/index.html [hunk 5]
```diff
@@ -709,7 +709,11 @@
     <label for="locale">{{ _("Language:") }}</label>
     <select name="locale" id="locale">
     {% for locale in locales %}
-        <option value="{{ locale.language }}">
+        {% if locale.language == env['locale'] %}
+            <option value="{{ locale.language }}" selected="selected">
+        {% else %}
+            <option value="{{ locale.language }}">
+        {% endif %}
```
Comment (rdmark) [inline python/web/src/templates/index.html:716]: I thought about adding this logic while implementing localization, but decided against it at the time. It doesn't hurt though. Maybe you were sick of seeing "Deutsch" being selected all the time? ;)

File: python/web/src/web.py [hunk 6]
```diff
@@ -168,16 +168,7 @@ def get_locale():
...
-        language = session["language"]
-    except KeyError:
-        language = ""
-        logging.info("The default locale could not be detected. Falling back to English.")
-    if language:
-        return language
-    # Hardcoded fallback to "en" when the user agent does not send an accept-language header
-    language = request.accept_languages.best_match(LANGUAGES) or "en"
-    return language
+    return session.get("language", request.accept_languages.best_match(LANGUAGES) or "en")
```
Comment (rdmark) [inline python/web/src/web.py]: Definitely neater logic. Did you test this will vintage browsers? I recall running into corner cases with Netscape 3/4 last year which is when I added the elaborate logic above.

File: python/web/src/web.py [hunk 7]
```diff
@@ -1381,5 +1408,4 @@ def detect_locale():
         except KeyboardInterrupt:
             pass
     else:
-        print("Serving rascsi-web...")
```
Comment (rdmark) [inline python/web/src/web.py:1384]: Why did you remove this? It is marginally helpful to know when the startup sequence has reached the bjoern step.

File: python/web/start.sh [hunk 8]
```diff
@@ -121,13 +121,14 @@ while [ "$1" != "" ]; do
 done
 
 PYTHON_COMMON_PATH=$(dirname $PWD)/common/src
-echo "Starting web server for RaSCSI Web Interface..."
 export PYTHONPATH=$PWD/src:${PYTHON_COMMON_PATH}
 cd src
 
 if [[ $ARG_DEV_MODE ]]; then
+    echo "Starting web UI (auto-reload) ..."
```
Comment (rdmark) [inline python/web/start.sh:128]: "auto-reload" is equivalent to "dev mode", right? Any particular reason for not using the same terminology?

**Review by rdmark — APPROVED** [review]

File: docker/README.md (hunk 2 above)
Comment (nucleogenic) [reply docker/README.md:46]: Definitely, but I think in a separate PR, once the rename-a-thon begins?

File: .dockerignore (hunk 1 above)
Comment (nucleogenic) [reply .dockerignore:5]: This is actually just `rascsi/rascsi_wrapper.sh` in the new backend dir; more places to update/test once the script is renamed.

File: python/web/src/templates/index.html (hunk 5 above)
Comment (nucleogenic) [reply python/web/src/templates/index.html:716]: Pretty much! I don't think there's anywhere to see the current locale, otherwise?

File: python/web/src/web.py (hunk 6 above)
Comment (nucleogenic) [reply python/web/src/web.py]: See updated commit, which restores the log. Will test in Netscape to confirm no regressions.

File: python/web/src/web.py (hunk 7 above)
Comment (nucleogenic) [reply python/web/src/web.py:1384]: Re-added generic message higher up, see latest commit. Suggestions for wording welcome (or feel free to change in a future change set!)

File: python/web/start.sh (hunk 8 above)
Comment (nucleogenic) [reply python/web/start.sh]: Auto-reload is certainly a consequence of the dev mode flag. Swapped it back to dev mode for consistency in latest commit.

File: python/web/src/web.py (hunk 6 above)
Comment (nucleogenic) [reply python/web/src/web.py]: Okay, so I only tested in Netscape 2.x, as it's what I had available, but works as expected.

I did however find an issue with `browser_supports_modern_themes()` which I've pushed a commit to fix!

image
